File: Leetcode/Linked List/234.palindrome-linked-list.java
/*
Given a singly linked list, determine if it is a palindrome.

Example 1:

Input: 1->2
Output: false

Example 2:

Input: 1->2->2->1
Output: true

Follow up:
Could you do it in O(n) time and O(1) space?

******************************************************************************************Solution*****************************************************************************************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    //The idea is to split the list into parts, reverse the 2nd half and then compare with 1st half 
    public boolean isPalindrome(ListNode head) 
    {
        ListNode temp = head, mid = head;
        
        while(temp != null && temp.next != null)
        {
            mid = mid.next;
            temp = temp.next.next;
        }
        
        ListNode headSecondHalf = null;
        
        //when number of nodes is even, mid node would be the head of 2nd half of the list
        if(temp == null) 
            headSecondHalf = mid;
        //when number of nodes is odd, next node of mid node would be the head of 2nd half of the list
        else 
            headSecondHalf = mid.next;
        
        headSecondHalf = reverse(headSecondHalf);
        
        return compare(head, mid, headSecondHalf);
    }
    
    //to reverse the 2nd half of the list
    ListNode reverse(ListNode head)
    {
        ListNode prev = null, current = head, next;
        
        while(current != null)
        {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        
        return prev;
    }
    
    //to compare the 1st half of the list is matching with 2nd half or not
    boolean compare(ListNode headFirstHalf, ListNode endFirstHalf, ListNode headSecondHalf)
    {
        while(headFirstHalf != endFirstHalf && headSecondHalf != null)
        {
            if(headFirstHalf.val != headSecondHalf.val)
                return false;
            
            headFirstHalf = headFirstHalf.next;
            headSecondHalf = headSecondHalf.next;
        }
        
        return true;
    }
}

File: DSA-JavaScript/Linked List/234.palindrome-linked-list.js
/*

Given the head of a singly linked list, return true if it is a palindrome or false otherwise.


Example 1:

Input: head = [1,2,2,1]
Output: true

Example 2:

Input: head = [1,2]
Output: false
 

Constraints:

The number of nodes in the list is in the range [1, 10^5].
0 <= Node.val <= 9

**************************************************************Solution**************************************************************/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */

/* the idea is to split the list into 2 parts, reverse the 2nd part of the list and then compare reversed 2nd part of the 
list with the 1st part to chcek whether the list is palindrome or not */

const isPalindrome = function(head) {
    const middleNodeOfTheList = findMiddleNode(head);
    
    const headOfThe2ndPartOfTheList = middleNodeOfTheList.next;    
    const headOfTheReversed2ndPartOfTheList = reverseList(headOfThe2ndPartOfTheList);
    
    return compare2PartsOfTheList(head, headOfTheReversed2ndPartOfTheList);
};

const findMiddleNode = (head) => {
    let fastPointer = head, slowPointer = head;
    
    while(fastPointer.next && fastPointer.next.next){
        slowPointer = slowPointer.next;
        fastPointer = fastPointer.next.next;
    }
    
    return slowPointer;
}

const reverseList = (head) => {
    let currentNode = head, nextNode, previousNode = null;
    
    while(currentNode){
        nextNode = currentNode.next;
    
        currentNode.next = previousNode;        
        previousNode = currentNode;    
        currentNode = nextNode;
    }    
    
    return previousNode;
}

const compare2PartsOfTheList = (headOf1stPart, headOf2ndPart) => {
    // 2nd part of the list would have null to mark the end of the list
    while(headOf2ndPart){
        if(headOf1stPart.val !== headOf2ndPart.val){
            return false
        }
        
        headOf1stPart = headOf1stPart.next;
        headOf2ndPart = headOf2ndPart.next;
    }   
    
    return true;
}
